nomad/node_endpoint.go

```diff
@@ -687,10 +697,22 @@ func (n *Node) GetClientAllocs(args *structs.NodeSpecificRequest,
 			}
 
 			reply.Allocs = make(map[string]uint64)
+			reply.MigrateTokens = make(map[string]string)
 			// Setup the output
 			if len(allocs) != 0 {
 				for _, alloc := range allocs {
 					reply.Allocs[alloc.ID] = alloc.AllocModifyIndex
+
+					allocNode, err := state.NodeByID(ws, args.NodeID)
```

Comment:
You actually want to be using the secret id of the previous node.

```
if alloc.ShouldMigrate() {
  // Lookup up previous allocation.
  // If previous allocation is on a different node then lookup that node
  // Generate the token.
```
Also modify ShouldMigrate to check that `alloc.PreviousAllocation != ""`

This way we generate tokens only when necessary too ^